### PR TITLE
Add type declarations where backwards-compatible

### LINKED ISSUE #8538: Add backwards-compatible type declarations
We should grep for:
- `private function`
- `final protected function`
- `final class`

And see if we can add type declarations.
Since the phpdoc might be inaccurate, this is a risky move, and as such should IMO target 2.9.x

### PR DESCRIPTION
**Pull request by greg0ire**
This includes:
- private methods
- return type declarations of final protected methods
- return type declarations of public and protected methods of final classes

Parameter type declarations are a more delicate matter and should
probably be handled separately to make it easier to catch issues during
code review.

Type declarations can be more trusted than simple phpdoc when it
comes to static analysis, having them means we can infer the phpdoc of
calling methods with confidence.

Note that it seems that some of the phpdoc I initially inferred these
declarations from were apparently wrong, in particular some mentioning
`Doctrine\Dbal\Statement` when was is really passed around is
`Doctrine\Dbal\Driver\Statement`.

Closes #8538

### CODE REVIEW

File: lib/Doctrine/ORM/Internal/Hydration/ArrayHydrator.php [hunk 1]
```diff
@@ -261,13 +261,14 @@ protected function hydrateRowData(array $row, array &$result)
...
      * @param bool|int $index    Index of the element in the collection.
-     * @param string   $dqlAlias
      * @param bool     $oneToOne Whether it is a single-valued association or not.
-     *
-     * @return void
      */
-    private function updateResultPointer(array &$coll, $index, $dqlAlias, $oneToOne)
-    {
+    private function updateResultPointer(
+        array &$coll,
```
Comment (simonberger) [inline lib/Doctrine/ORM/Internal/Hydration/ArrayHydrator.php]: `null` is handled in the code for parameter `$coll`.

File: lib/Doctrine/ORM/PersistentCollection.php [hunk 2]
```diff
@@ -292,17 +277,15 @@ static function ($a, $b): int {
      *
      * @psalm-return array<string, mixed>
      */
-    public function getMapping()
+    public function getMapping(): array
```
Comment (simonberger) [inline lib/Doctrine/ORM/PersistentCollection.php]: Result can be null. Property var attribute needs a change as well.

**Comment by greg0ire** [comment]
Thanks for the careful review @simonberger , much appreciated 👍

**Review by simonberger — DISMISSED** [review]

File: lib/Doctrine/ORM/Cache/DefaultQueryCache.php [hunk 3]
```diff
@@ -344,11 +344,9 @@ public function put(QueryCacheKey $key, ResultSetMapping $rsm, $result, array $h
      * @param array<string,mixed> $assoc
      * @param mixed               $assocValue
      *
-     * @return mixed[]|null
```
Comment (BackEndTea) [inline lib/Doctrine/ORM/Cache/DefaultQueryCache.php:347]: I'd keep the annotation here. That at least makes it clear the return type wasn't forgotten here

File: lib/Doctrine/ORM/Cache/Persister/Entity/NonStrictReadWriteCachedEntityPersister.php [hunk 4]
```diff
@@ -98,13 +98,7 @@ public function update($entity)
...
     }
 
-    /**
-     * @param object $entity
-     * @param bool   $isChanged
-     *
-     * @return bool
-     */
-    private function updateCache($entity, $isChanged)
+    private function updateCache(object $entity, bool $isChanged): ?bool
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Cache/Persister/Entity/NonStrictReadWriteCachedEntityPersister.php]: rather than changing it from docblock version, i would prefer fixing the code here: `$isChanged = $isChanged || $cached;`

File: lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php [hunk 5]
```diff
@@ -273,13 +272,11 @@ private function getEntity(array $data, $dqlAlias)
     }
 
     /**
-     * @param string $className
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:275]: isn't this a class-string? should be marked as such, if so

File: lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php [hunk 6]
```diff
@@ -384,12 +382,8 @@ private function addDefaultDiscriminatorMap(ClassMetadata $class)
 
     /**
      * Gets the lower-case short name of a class.
-     *
-     * @param string $className
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php:393]: class-string?

File: lib/Doctrine/ORM/Mapping/Driver/DatabaseDriver.php [hunk 7]
```diff
@@ -526,12 +522,8 @@ private function getTablePrimaryKeys(Table $table)
 
     /**
      * Returns the mapped class name for a table if it exists. Otherwise return "classified" version.
-     *
-     * @param string $tableName
-     *
-     * @return string
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Mapping/Driver/DatabaseDriver.php:537]: class-string?

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php [hunk 8]
```diff
@@ -821,11 +819,9 @@ private function columnToArray(SimpleXMLElement $fieldMapping)
     /**
      * Parse / Normalize the cache configuration
      *
-     * @return mixed[]
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php:878]: Sometimes you remove these, sometimes don't, like in yamldriver or PersistentCollection. So which way is it, what's the policy?

File: lib/Doctrine/ORM/NativeQuery.php [hunk 9]
```diff
@@ -60,7 +60,7 @@ public function getSQL()
     }
 
     /**
-     * {@inheritdoc}
+     * {@inheritDoc}
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/NativeQuery.php]: what was your motivation to change these? I fuzzily remember phpcs understood lowercase variant only

File: lib/Doctrine/ORM/PersistentCollection.php [hunk 10]
```diff
@@ -143,18 +141,13 @@ public function setOwner($entity, array $assoc)
...
+    public function getOwner(): ?object
     {
         return $this->owner;
     }
 
-    /**
-     * @return Mapping\ClassMetadata
-     */
-    public function getTypeClass()
+    public function getTypeClass(): Mapping\ClassMetadata
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/PersistentCollection.php]: This one is dangerous. We don't enforce type in constructor and there also is ClassMetadataInfo

File: lib/Doctrine/ORM/PersistentCollection.php [hunk 11]
```diff
@@ -348,20 +329,16 @@ public function setDirty($dirty)
...
         $this->initialized = $bool;
     }
 
     /**
      * {@inheritdoc}
-     *
-     * @return object
      */
-    public function remove($key)
+    public function remove($key): object
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/PersistentCollection.php]: should be `?object`. It calls parent::remove, ArrayCollection returns null in some cases and interface also says T|null.

please also take care of references to this method, like offsetUnset

File: lib/Doctrine/ORM/Query.php [hunk 12]
```diff
@@ -279,7 +277,7 @@ private function parse()
     }
 
     /**
-     * {@inheritdoc}
+     * {@inheritDoc}
      */
     protected function _doExecute()
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Query.php]: return type here should probably be added

File: lib/Doctrine/ORM/Query.php [hunk 13]
```diff
@@ -523,10 +521,8 @@ public function setQueryCacheLifetime($timeToLive): self
 
     /**
      * Retrieves the lifetime of resultset cache.
-     *
-     * @return int
      */
-    public function getQueryCacheLifetime()
+    public function getQueryCacheLifetime(): int
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Query.php]: Not sure why SA isn't complaining here. `queryCacheTTL` prop is uninitializd so I think it may return null

File: lib/Doctrine/ORM/Query/ResultSetMappingBuilder.php [hunk 14]
```diff
@@ -223,15 +225,13 @@ private function getColumnAlias($columnName, $mode, array $customRenameColumns)
      *
      * This depends on the renaming mode selected by the user.
      *
-     * @param string $className
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Query/ResultSetMappingBuilder.php:222]: class-string?

File: lib/Doctrine/ORM/Query/ResultSetMappingBuilder.php [hunk 15]
```diff
@@ -223,15 +225,13 @@ private function getColumnAlias($columnName, $mode, array $customRenameColumns)
...
-     * @param int    $mode
-     *
      * @return string[]
      *
+     * @psalm-param self::COLUMN_RENAMING_* $mode
      * @psalm-param array<string, string> $customRenameColumns
      * @psalm-return array<array-key, string>
      */
-    private function getColumnAliasMap($className, $mode, array $customRenameColumns)
+    private function getColumnAliasMap(string $className, $mode, array $customRenameColumns): array
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Query/ResultSetMappingBuilder.php]: `int $mode`? Either that or docblock should be fixed, shouldn't it?

File: lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php [hunk 16]
```diff
@@ -95,24 +95,16 @@ public function generateEntityRepositoryClass($fullClassName)
 
     /**
      * Generates the namespace, if class do not have namespace, return empty string instead.
-     *
-     * @param string $fullClassName
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php:99]: class-string

File: lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php [hunk 17]
```diff
@@ -95,24 +95,16 @@ public function generateEntityRepositoryClass($fullClassName)
...
-    private function getClassNamespace($fullClassName)
+    private function getClassNamespace(string $fullClassName): string
     {
         return substr($fullClassName, 0, strrpos($fullClassName, '\\'));
     }
 
     /**
      * Generates the class name
-     *
-     * @param string $fullClassName
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php:111]: class-string

File: lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php [hunk 18]
```diff
@@ -129,22 +121,15 @@ private function generateClassName($fullClassName)
      * Generates the namespace statement, if class do not have namespace, return empty string instead.
      *
      * @param string $fullClassName The full repository class name.
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php]: class-string

File: lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php [hunk 19]
```diff
@@ -129,22 +121,15 @@ private function generateClassName($fullClassName)
...
-    private function generateEntityRepositoryNamespace($fullClassName)
+    private function generateEntityRepositoryNamespace(string $fullClassName): string
     {
         $namespace = $this->getClassNamespace($fullClassName);
 
         return $namespace ? 'namespace ' . $namespace . ';' : '';
     }
 
-    /**
-     * @param string $fullClassName
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php:143]: class-string

File: lib/Doctrine/ORM/Tools/Pagination/Paginator.php [hunk 20]
```diff
@@ -219,10 +208,8 @@ private function useOutputWalker(Query $query)
 
     /**
      * Appends a custom tree walker to the tree walkers hint.
-     *
-     * @param string $walkerClass
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Tools/Pagination/Paginator.php:224]: class-string

File: lib/Doctrine/ORM/Tools/ResolveTargetEntityListener.php [hunk 21]
```diff
@@ -110,12 +109,9 @@ public function loadClassMetadata(LoadClassMetadataEventArgs $args)
...
 
     /**
-     * @param ClassMetadataInfo $classMetadata
-     * @param mixed[]           $mapping
-     *
-     * @return void
+     * @param mixed[] $mapping
      */
-    private function remapAssociation($classMetadata, $mapping)
+    private function remapAssociation(ClassMetadata $classMetadata, $mapping): void
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Tools/ResolveTargetEntityListener.php]: any reason why not declaring type for $mapping?

**Review by ostrolucky — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/Query.php (hunk 13 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query.php]: I think that's because we are at a dangerously lenient level right now.

File: lib/Doctrine/ORM/Tools/ResolveTargetEntityListener.php (hunk 21 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Tools/ResolveTargetEntityListener.php]: That one slipped through :P

File: lib/Doctrine/ORM/NativeQuery.php (hunk 9 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/NativeQuery.php]: I don't remember, and we seem to be using a mix of both in the project :sweat: .

Will look into it later.

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php (hunk 8 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php:878]: It's unclear to me what's best, will give it more thought later.

File: lib/Doctrine/ORM/Query.php (hunk 12 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query.php]: That one causes so many test failures :scream:

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php (hunk 8 above)
Comment (ostrolucky) [reply lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php:878]: If I may, I slightly prefer keeping them. There might be tools that don't respect psalm vendor specific annotations and mixed[] makes it clear to communicate we didn't forget to specify the type.

File: lib/Doctrine/ORM/Query.php (hunk 12 above)
Comment (ostrolucky) [reply lib/Doctrine/ORM/Query.php]: So is it actually `mixed`?

File: lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php (hunk 8 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php:878]: Ok, I'll restore them then 👍

File: lib/Doctrine/ORM/Query.php (hunk 12 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Query.php]: Indeed, it's not just the tests, the sql executor hierarchy has inconsistent return type declarations. `MultiTableDeleteExecutor` declares `int`, so does `SingleTableDeleteUpdateExecutor`.

**Review by ostrolucky — COMMENTED** [review]
One comment remaining. Also, you are adding object type declarations, but these are being removed in https://github.com/doctrine/orm/pull/8613. That will cause mess of a conflict.

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php [hunk 22]
```diff
@@ -24,9 +24,9 @@
 use Doctrine\Common\Collections\Expr\Comparison;
 use Doctrine\Common\Util\ClassUtils;
 use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Driver\ResultStatement as DriverStatement;
 use Doctrine\DBAL\LockMode;
 use Doctrine\DBAL\Platforms\AbstractPlatform;
-use Doctrine\DBAL\Statement;
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:29]: Unsure if we should wait for https://github.com/doctrine/dbal/pull/4596 and revert this change in use statements. I think that change might be good because more forward compatible.

Comment (ostrolucky) [reply lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:29]: I'll leave you decide. But to me, such change seems a bit unrelated to this PR

**Comment by greg0ire** [comment]
I will wait until #8613 is resolved

**Review by ostrolucky — DISMISSED** [review]

**Comment by greg0ire** [comment]
The 7.1 PR has been merged, let's rebase to be 100% sure it still works
